quantum/impl/quantum_read_write_mutex_impl.h

```diff
@@ -143,35 +175,40 @@ inline
 ReadWriteMutex::ReadGuard::ReadGuard(ReadWriteMutex& lock,
                                      ReadWriteMutex::TryToLock) :
     _mutex(&lock),
-    _ownsLock(_mutex->tryLockRead())
+    _ownsLock(_mutex->tryLockRead()),
+    _isUpgraded(false)
 {
 }
 
 inline
 ReadWriteMutex::ReadGuard::ReadGuard(ReadWriteMutex& lock,
                                      ReadWriteMutex::AdoptLock) :
     _mutex(&lock),
-    _ownsLock(true)
+    _ownsLock(lock.isReadLocked()),
```

Comment:
This looks wrong; if we adopt an upgraded lock, _ownsLock will be false but _isUpgraded will be true. Then we won't unlock in the destructor.